### PR TITLE
Refactor seed

### PR DESCRIPTION
**Pull request by pedrocodacy**

### CODE REVIEW

File: src/test/scala/codacy/docker/api/duplication/DockerDuplicationSpecs.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            status must beEqualTo(0)
+            ()
+          }
+        }
+
+      //when
+      dockerDuplication.main(Array.empty)
+
+      //then
+      Json.parse(outContent.toString) mustEqual Json.toJson(
```
Comment (rtfpessoa) [inline src/test/scala/codacy/docker/api/duplication/DockerDuplicationSpecs.scala]: Same as in the other json test

File: src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.io.PrintStream
+
+import codacy.docker.api.duplication.DuplicationClone
+import play.api.libs.json.Json
+
+import scala.util.Try
+
+class ResultsPrinter(resultsStream: PrintStream = Console.out, logStream: PrintStream = Console.err) {
+
+  lazy val isDebug: Boolean = Option(System.getProperty("debug")).flatMap { rawDebug =>
```
Comment (rtfpessoa) [inline src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala]: read this from the environment thought a parameter for it to be more testable.

File: src/main/scala/codacy/docker/api/utils/FileHelper.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  def listAllFiles(path: String): List[File] = {
+    listAllFiles(Paths.get(path))
+  }
+
+  def listAllFiles(path: Path): List[File] = {
+    recursiveListFiles(path.toFile)
+  }
+
+  private def recursiveListFiles(file: File): List[File] = {
+    val these = file.listFiles
```
Comment (rtfpessoa) [inline src/main/scala/codacy/docker/api/utils/FileHelper.scala]: Eventually we should replace this file ops by better files.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      - *restore_build
+      - run:
+          name: Install AWS CLI
+          command: |
+            sudo apt -y update
+            sudo apt -y install python3-pip
+            sudo python3 -m pip install awscli
+      - run:
+          name: Setup AWS Credentials
+          command: ./scripts/setup-aws-credentials.sh
```
Comment (rtfpessoa) [inline .circleci/config.yml]: Did you add this script to the repo?

**Comment by rtfpessoa** [comment]
Also check the README @bmbferreira did for metrics and update this one.

File: src/main/scala/codacy/docker/api/duplication/package.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        JsObject(m.flatMap {
+          case (k, v: DuplicationConfigurationValue) => Option(k.value -> v.value)
+          case _                                     => Option.empty[(String, JsValue)]
+        })
+      })
+
+  implicit val dupCloneFileFmt: OFormat[DuplicationCloneFile] = Json.format[DuplicationCloneFile]
+  implicit val dupCloneFmt: OFormat[DuplicationClone] = Json.format[DuplicationClone]
+  implicit val dupCfgFmt: OFormat[DuplicationConfiguration] = Json.format[DuplicationConfiguration]
+  implicit val codacyCfgFmt: OFormat[CodacyConfiguration] = Json.format[CodacyConfiguration]
```
Comment (xplosunn) [inline src/main/scala/codacy/docker/api/duplication/package.scala]: Why are some of the vals above lazy and some aren't?

File: src/main/scala/codacy/docker/api/duplication/package.scala [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+}
+
+private[this] case class DuplicationConfigurationValue(value: JsValue) extends AnyVal with Value
```
Comment (xplosunn) [inline src/main/scala/codacy/docker/api/duplication/package.scala]: Move this to the top of the file and inside the package object.

File: src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      logStream.println(message)
+    }
+  }
+
+  def logStackTrace(error: Throwable): Unit = {
+    error.printStackTrace(logStream)
+  }
+
+  def printResults(results: Seq[DuplicationClone], sourcePath: String): Unit = {
+
```
Comment (xplosunn) [inline src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala]: Unnecessary empty line

File: src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    results.foreach { result =>
+      resultsStream.println(Json.stringify(Json.toJson(result.copy(files = result.files.map(file =>
+        file.copy(filePath = stripSourcePath(file.filePath, sourcePath)))))))
```
Comment (xplosunn) [inline src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala]: There's too much nesting here. Move something out into a val or function to improve readability.

File: src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    }
+  }
+
+  private def stripSourcePath(fullPath: String, sourcePath: String): String = {
```
Comment (xplosunn) [inline src/main/scala/codacy/docker/api/utils/ResultsPrinter.scala]: This function strikes me as unnecessary. It's just a proxy.

File: src/main/scala/codacy/docker/api/utils/Timeout.scala [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.{Timer, TimerTask}
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Future, Promise}
+import scala.util.Try
+
+trait Timeout {
+
+  def initTimeout[T](delay: Duration)(onTimeout: => T): Future[T] = {
```
Comment (xplosunn) [inline src/main/scala/codacy/docker/api/utils/Timeout.scala]: "initTimeout" sounds weird to me. I'd rather have it named "onTimeout", like you did for the parameter.

File: src/test/scala/codacy/docker/api/utils/TimeoutSpecs.scala [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, TimeoutException}
+
+class TimeoutSpecs extends Specification with Timeout {
+
+  "Timeout" should {
+    "should throw exception" in {
+      val f = initTimeout(100.seconds) {
+        failure("It should fail because a timeoutException wasn't thrown.")
```
Comment (xplosunn) [inline src/test/scala/codacy/docker/api/utils/TimeoutSpecs.scala]: Change the text to "A scala.concurrent.TimeoutException" was expected to make the test easier to understand.

File: src/test/scala/codacy/docker/api/utils/TimeoutSpecs.scala [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+      }
+
+      Await.result(f, 1.second) must throwA[TimeoutException]
+    }
+
+    "shouldn't throw exception" in {
+      val f = initTimeout(1.seconds) {
+        success("The delay didn't throw a TimeoutException exception")
```
Comment (xplosunn) [inline src/test/scala/codacy/docker/api/utils/TimeoutSpecs.scala]: It's not the delay that throws the TimeoutException.

File: src/main/scala/codacy/docker/api/duplication/DockerDuplication.scala [hunk 13]
```diff
@@ -0,0 +1,48 @@
+package codacy.docker.api.duplication
+
+import codacy.docker.api.Source.Directory
+import codacy.docker.api.utils.{Haltable, ResultsPrinter, Timeout}
+
+import scala.util.{Failure, Success, Try}
+
+class DockerDuplication(tool: DuplicationTool,
+                        environment: DockerDuplicationEnvironment = new DockerDuplicationEnvironment)(
+  printer: ResultsPrinter = new ResultsPrinter(isDebug = environment.isDebug))
```
[the remark this answers is not in the document]
Comment (pedrocodacy) [reply src/main/scala/codacy/docker/api/duplication/DockerDuplication.scala]: Necessary to be able to use the default initialization of environment to initialize the results printer

File: src/main/scala/codacy/docker/api/duplication/package.scala (hunk 6 above)
Comment (pedrocodacy) [reply src/main/scala/codacy/docker/api/duplication/package.scala]: Only moved to top.
From compiler:
[error]: .../codacy-duplication-scala-seed/src/main/scala/codacy/docker/api/duplication/package.scala:13:28: it is not recommended to define classes/objects inside of package objects.

File: src/main/scala/codacy/docker/api/duplication/package.scala (hunk 5 above)
Comment (pedrocodacy) [reply src/main/scala/codacy/docker/api/duplication/package.scala]: Made them all non-lazy

File: .circleci/config.yml (hunk 4 above)
Comment (pedrocodacy) [reply .circleci/config.yml]: done

**Pull request merged**